### ISSUE TITLE
Does CatBoostRegressor minimize RMSE?

### ISSUE THREAD

**Issue by gitmario** [catboost/catboost#627]
Problem: {CatBoostRegressor unexpected output}
```
import pandas as pd
from catboost import CatBoostRegressor

df_train = pd.DataFrame({'segment':['A','A','A',
                                    'B','B','B',
                                    'C','C','C',],
                         'y':[1, 2, 3,
                              10, 20, 30,
                              0.1, 0.2, 0.3]})

df_test = pd.DataFrame({'segment':['A','A',
                                   'B','B',
                                   'C','C']})

ml_model=CatBoostRegressor()

ml_model.fit(X=df_train.drop(columns=['y']),
             y=df_train['y'],
             cat_features=[0],
             verbose=False)

ml_model.predict(df_test)
```
Output:
```
array([ 2.3269586 ,  2.3269586 , 28.23650746, 28.23650746,  0.27892251, 0.27892251])
```
Expected output:
```
array([ 2.0 ,  2.0 , 20.0, 20.0,  0.2, 0.2])
```
catboost version: {catboost==0.11.2}
Operating System: {macOS}

**Comment by gitmario** [comment]
By the way, this only happens with categorical encoding:
```
import pandas as pd
from catboost import CatBoostRegressor

df_train = pd.DataFrame({'segment':[1,1,1,
                                    2,2,2,
                                    3,3,3],
                         'y':[1, 2, 3,
                              10, 20, 30,
                              0.1, 0.2, 0.3]})

df_test = pd.DataFrame({'segment':[1,1,
                                   2,2,
                                   3,3]})

ml_model=CatBoostRegressor()

ml_model.fit(X=df_train.drop(columns=['y']),
             y=df_train['y'],
             # cat_features=[0],
             verbose=False)

ml_model.predict(df_test)
```
Output:
```
array([ 2.00353531,  2.00353531, 19.99226603, 19.99226603,  0.20556117,
        0.20556117])
```

**Comment by david-waterworth** [comment]
Seems similar to my issue - https://github.com/catboost/catboost/issues/622

The CatBoostRegressor doesn't minimise the mse for a trivial example (with no categorical data, my actual problem does have categories). Also at each boosting step the residuals don't have zero mean - gradient boosting with mse normally starts with a constant model and at each should reduce the error but the mean should be zero?

Not sure if this has something to do with it? https://arxiv.org/abs/1706.09516

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
Hi @gitmario -- I was just passing by and noticed your question.
 
By default, CatBoost replaces cat features with numeric features.
Category X at row N is replaced with something like count(X in [0, N-1]) / N -- please refer to https://catboost.ai/docs/concepts/algorithm-main-stages_cat-to-numberic.html for details.
Thus, by default, the first A and the second A differ for CatBoost.
To make them exactly same, please enable one-hot encoding by setting one_hot_max_size=number of different categories, e.g. ml_model=CatBoostRegressor(one_hot_max_size=3) in your example.
With this setting, we get the expected output.
